experiment/needs-rebase/plugin/plugin.go

```diff
@@ -0,0 +1,249 @@
+	sleep(time.Second * 5)
+
+	org := pre.Repo.Owner.Login
+	repo := pre.Repo.Name
+	number := pre.Number
+	sha := pre.PullRequest.Head.SHA
+
+	mergeable, err := ghc.IsMergeable(org, repo, number, sha)
+	if err != nil {
+		return err
+	}
+	labels, err := ghc.GetIssueLabels(org, repo, number)
+	if err != nil {
+		return err
+	}
+	hasLabel := github.HasLabel(needsRebaseLabel, labels)
+
+	return takeAction(log, ghc, org, repo, number, pre.PullRequest.User.Login, hasLabel, mergeable)
+}
+
+func HandleAll(log *logrus.Entry, ghc githubClient, config *plugins.Configuration) error {
+	orgs, repos := config.EnabledReposForPlugin(pluginName)
+	var buf bytes.Buffer
+	fmt.Fprint(&buf, "is:pr is:open")
+	for _, org := range orgs {
+		fmt.Fprintf(&buf, " org:\"%s\"", org)
+	}
+	for _, repo := range repos {
+		fmt.Fprintf(&buf, " repo:\"%s\"", repo)
+	}
+	prs, err := search(context.Background(), log, ghc, buf.String())
+	if err != nil {
+		return err
+	}
+
+	for _, pr := range prs {
+		// Skip PRs that are calculating mergeability. They will be updated by event or next loop.
+		if pr.Mergeable == githubql.MergeableStateUnknown {
+			continue
+		}
+		org := string(pr.Repository.Owner.Login)
+		repo := string(pr.Repository.Name)
+		num := int(pr.Number)
+		l := log.WithFields(logrus.Fields{
+			"org":  org,
+			"repo": repo,
+			"pr":   num,
+		})
+		hasLabel := false
+		for _, node := range pr.Labels.Nodes {
+			if node.Label.Name == needsRebaseLabel {
+				hasLabel = true
+				break
+			}
+		}
+		err := takeAction(
+			l,
+			ghc,
+			org,
+			repo,
+			num,
+			string(pr.Author.Login),
+			hasLabel,
+			pr.Mergeable == githubql.MergeableStateMergeable,
+		)
+		if err != nil {
+			l.WithError(err).Error("Error handling PR.")
+		}
+	}
+	return nil
+}
+
+func takeAction(log *logrus.Entry, ghc githubClient, org, repo string, num int, author string, hasLabel, mergeable bool) error {
+	if !mergeable && !hasLabel {
+		if err := ghc.AddLabel(org, repo, num, needsRebaseLabel); err != nil {
+			log.WithError(err).Errorf("Failed to add %q label.", needsRebaseLabel)
+		}
+		msg := plugins.FormatSimpleResponse(author, needsRebaseMessage)
+		return ghc.CreateComment(org, repo, num, msg)
+	} else if mergeable && hasLabel {
+		// remove label and prune comment
+		if err := ghc.RemoveLabel(org, repo, num, needsRebaseLabel); err != nil {
+			log.WithError(err).Errorf("Failed to remove %q label.", needsRebaseLabel)
+		}
+		botName, err := ghc.BotName()
+		if err != nil {
+			return err
+		}
+		return ghc.DeleteStaleComments(org, repo, num, nil, shouldPrune(botName))
+	}
+	return nil
+}
+
+func shouldPrune(botName string) func(github.IssueComment) bool {
+	return func(ic github.IssueComment) bool {
+		return github.NormLogin(botName) == github.NormLogin(ic.User.Login) &&
+			strings.Contains(ic.Body, needsRebaseMessage)
+	}
+}
+
+func search(ctx context.Context, log *logrus.Entry, ghc githubClient, q string) ([]pullRequest, error) {
```

Comment:
Maybe we can share code between this and tide in a follow-up.

Reply:
Yeah that would be an improvement. GraphQL + golang makes it a bit tricky because the `pullRequest` struct will be different in each use case because we only fetch the fields we need for the use case. I think we could work around that with interfaces though.